### DISCUSSION TITLE
PDF not opening on a separate window.

### DISCUSSION THREAD

**Discussion by terry2026** [cefsharp/CefSharp#4843] [General]
We did a CEFSharp jump for version 89 to version 117. The upgrade went fine just a couple of code modifications nothing major. The problem is since the upgrade, the PDF documents do not open in a separate window. In version 89 was working as expected, we have the Disable-pdf-extension and we implemented the downloadhandler. But since the upgrade I click on the PDF link it doesn't do anything.

**Comment by amaitland** [comment]
Please no images of code. See https://idownvotedbecau.se/imageofcode

> We did a CEFSharp jump for version 89 to version 117

How does the current supported version behave? (M125 at time of writing).

> But since the upgrade I click on the PDF link it doesn't do anything.

Did you check the log file?

**Reply by terry2026** [reply]
Thank you for your reply.

Sorry about the images.
```
var settings = new CefSettings();
            settings.CefCommandLineArgs.Add("disable-gpu", "1"); // Disable GPU acceleration
            settings.CefCommandLineArgs.Add("disable-gpu-vsync", "1"); //Disable GPU vsync
            settings.CefCommandLineArgs.Add("allow-running-insecure-content", "1"); //Enables mixed content on the start page
            settings.CefCommandLineArgs.Add("force-renderer-accessibility", "true"); //Enables UI automation
            settings.CefCommandLineArgs.Add("disable-pdf-extension"); //Disable PDF plugin to download the file instead
            Cef.Initialize(settings);
 /// <summary>
        /// Called before a download begins.
        /// </summary>
        /// <param name="chromiumWebBrowser"></param>
        /// <param name="browser"></param>
        /// <param name="downloadItem"></param>
        /// <param name="callback"></param>
        void IDownloadHandler.OnBeforeDownload(IWebBrowser chromiumWebBrowser, IBrowser browser, DownloadItem downloadItem, IBeforeDownloadCallback callback)
        {
            if (!callback.IsDisposed)
            {
                using (callback)
                {
                    Process.Start(downloadItem.Url);
                    callback.Continue(downloadItem.SuggestedFileName, false);
                }
            }
        }

        /// <summary>
        /// Called when a download's status or progress information has been updated. This
        ///     may be called multiple times before and after
        ///     CefSharp.IDownloadHandler.OnBeforeDownload(CefSharp.IWebBrowser,CefSharp.IBrowser,CefSharp.DownloadItem,CefSharp.IBeforeDownloadCallback).
        /// </summary>
        /// <param name="chromiumWebBrowser"></param>
        /// <param name="browser"></param>
        /// <param name="downloadItem"></param>
        /// <param name="callback"></param>
        /// Not returning anything. We don't need to monitor any updates to the download.
        void IDownloadHandler.OnDownloadUpdated(IWebBrowser chromiumWebBrowser, IBrowser browser, DownloadItem downloadItem, IDownloadItemCallback callback)
        {
            // do nothing for now
        }
  bool IRequestHandler.OnBeforeBrowse(IWebBrowser browserControl, IBrowser browser, IFrame frame, IRequest request, bool userGesture, bool isRedirect)
        {
            bool interruptBrowsing = false;
            Uri url;
            if (Uri.TryCreate(request.Url, UriKind.RelativeOrAbsolute, out url))
            {
                switch (url.Scheme.ToLower())
                {
                    case "chrome-devtools":
                        break;

                    case "mailto":
                        Process.Start(url.ToString()); // launch os registered email client
                        interruptBrowsing = true;
                        break;

                    default:
                        if (frame.IsMain) // dont open new tabs for links in sub frames
                        {
                            var args = new HtmlBrowserNavigateEventArgs(request.Url, null, isRedirect);
                            InvokeOnUIThread(() => this.Navigate?.Invoke(this, args));
                            interruptBrowsing = args.Cancel;
                        }
                        break;
                }
            }
            return interruptBrowsing;
        }
```
I upgraded to the latest version and it behaves the same way.
No messages in the log. Another thing that I did was modifying the minimal example and implement the hadlers the same way as an my code and the minimal example works. The only difference is that, the PDF request is in a subframe not in the main frame. Not very familiar with CEFsharp and I don't know if that makes a difference.

**Comment by terry2026** [comment]
I found that if I modify the code to the OnBeforeBrowse function  
default:
                        if (frame.IsMain) // dont open new tabs for links in sub frames
                        {
                            var args = new HtmlBrowserNavigateEventArgs(request.Url, null, isRedirect);
                            InvokeOnUIThread(() => this.Navigate?.Invoke(this, args));
                            interruptBrowsing = args.Cancel;
                        }
                        else
                        {
                            if (url.ToString().Contains(".pdf"))
                            {
                                Process.Start(url.ToString());
                                interruptBrowsing = true;
                            }
                        }
                        break;

the pdf documents open in a separate browser. Any idea why is this neccesary since in previous version it was not needed? The main issue is that when the frame.IsMain its true and the pdf extensions are disabled, the download it's not processed, even the new event Downloadhandler.CanDownload does not fire.

**Reply by amaitland** [reply]
> Sorry about the images.

Please format the code to make it readable (you can edit your post). See https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

> Any idea why is this neccesary since in previous version it was not needed?

You've skipped 20+ versions, in `Chromium` terms that's a massive amount of change. They've likely refactored/changed something.

`CefSharp` is just a wrapper around the `Chromium Embedded Framework(CEF)`. If you can reproduce the issue with M126, then you'll need to raise it at https://github.com/chromiumembedded/cef/issues

There have been major changes in M126 (upstream changes that is), so please make sure to test before opening an issue on the CEF issue tracker.

**Reply by terry2026** [reply]
I upgraded to M126 and I can't really say if the issue is resolved, for some reason the disable-pdf-extension setting is not working as expected. It looks like the setting is being ignored and is not triggering the download of the PDF even of the downloadhanlder is implemented. On version 117 if the setting was "1", the PDF plugin was not being displayed, not in this version, the plugin is displayed regardless of the value of the disable-pdf-extension. Thank you.

**Reply by amaitland** [reply]
> CefSharp is just a wrapper around the Chromium Embedded Framework(CEF). If you can reproduce the issue with M126, then you'll need to raise it at https://github.com/chromiumembedded/cef/issues

Raise an issue on the CEF issue tracker
